### DISCUSSION TITLE
Avalonia localization

### DISCUSSION THREAD

**Discussion by MelnikovDaniil** [AvaloniaUI/Avalonia#7131] [Ideas]
What features does Avalonia have for localizing applications?

**Comment by robloo** [comment]
WPF always had sub-par localization compared to WinForms and UWP. What would it take to add an x:UID and resources system like in UWP to Avalonia? This would be the best way to do it in my opinion.

**Comment by maxkatz6** [comment]
You can use something like https://github.com/DotNetPlus/ReswPlus
But some kind of MarkupExtension or AttachedProperty should be implemented to work with it.
UWP x:Uid is fast to use, and maybe is nice to have in Avalonia, but also is quite limited.

Upd: it seems like ReswPlus doesn't support resx, but only resw files. So UWP only sadly.

**Comment by robloo** [comment]
@MelnikovDaniil , take a look over at #3139 where @braca shows an example of one method of localizing that was done in the WPF days -- writing your own markup extension as @Tirraon was saying.

```
    public class TranslateExtension : MarkupExtension
    {
        public TranslateExtension(string key)
        {
            this.Key = key;
        }

        public string Key { get; set; }

        public string FallBack { get; set; }

        public string Context { get; set; }

        public override object ProvideValue(IServiceProvider serviceProvider)
        {
            ....

            var binding = new BindingExtension($"[{keyToUse}]")
            {
                Mode = BindingMode.OneWay,
                Source = Translator.Instance,
                FallbackValue = fallback,
            };

            return binding.ProvideValue(serviceProvider);
        }
    }
```

**Comment by robloo** [comment]
@maxkatz6 Here is another issue that can be moved to discussions now.

---

In UWP I've gotten quite used to [x:Uid based localization](https://docs.microsoft.com/en-us/windows/uwp/app-resources/localize-strings-ui-manifest) which is similar to what was done originally with [Windows Forms](https://docs.microsoft.com/en-us/dotnet/core/extensions/localization). However, I've increasingly noticed a trend towards localization using XAML markup extensions as I mentioned above.

Some points:
 - [Markup Extension for Resource lookup](https://devblogs.microsoft.com/ifdef-windows/use-a-custom-resource-markup-extension-to-succeed-at-ui-string-globalization/) is a good, relatively new article outlining the benefits.
 - Microsoft is doing this for some of their internal apps now
 - Lunacy is doing this: https://github.com/AvaloniaUI/Avalonia/pull/6909#issuecomment-965797941
 - This methodology works with UWP, WPF and Avalonia (it was perfected originally with WPF due to lack of localization support built into the framework)

Really the only downside with markup extension based localization is it doesn't support all the property types that are not string. It's sometimes necessary to localize widths or margins for example. In practice most controls and layouts adjust perfectly fine to different text lengths though. In a decently-sized app I have never needed to localize anything other than text.

Overall, I am strongly considering moving my app over to this localization technique. I wonder if perhaps Avalonia should have a built-in extension to support this out of the box.

**Reply by maxkatz6** [reply]
> Really the only downside with markup extension based localization is it doesn't support all the property types that are not string.

Are you sure? I haven't tried it, but target type information should be available in the markup extension, as well as it can return object of any type. In the end, usual bindings are just another markup extension example.

**Reply by robloo** [reply]
No, not sure -- my point was more it isn't automatic. As mentioned I've only ever needed localized strings in XAML-based UIs so haven't played with other types since Windows Forms which is too long ago to speak with confidence. 

Of course almost anything is possible with a markup extension and C# so having type information would allow you to do the conversion just like a converter. My point was it isn't available automatically, out-of-the-box. Almost anything can be done with x:Bind and markup extensions though if you are creative enough.

There is a real opportunity to build an extension for localization in Avalonia itself I think.

**Comment by npolyak** [comment]
take a look at https://www.codeproject.com/Articles/5317972/Theming-and-Localization-Functionality-for-Multipl

**Reply by robloo** [reply]
Another great article! Really interesting ideas too. Muddies the water a bit for me and I'll have to reconsider a few things.
